src/operator/custom/custom-inl.h

```diff
@@ -96,7 +96,14 @@ class CustomOperator {
       bool prev_recording = Imperative::Get()->set_is_recording(recording);
       bool prev_training = Imperative::Get()->set_is_training(training);
 
-      func();
+      try {
+        func();
+      } catch (dmlc::Error& e) {
+        exception_ =
+            std::make_shared<std::exception_ptr>(std::current_exception());
+        ctx.async_on_complete();
+        return;
+      }
```

Comment:
I think we can solve both 1 and 2 this way: After func is called do wait_to_read on all elements in arrs. Then catch and save. Remove lines 104 and 105. In PushSync, check if exception is set and rethrow exception. Also catch it and call async_on_complete in pushsync. and return.

Something like the following:

```
Engine::Get()->PushSync(
     [=](RunContext rctx) {
    try {
    if (exception_) {
       std::rethrow_exception(exception_);
    }
    } catch(dmlc::Error& err) {
         ctx.async_on_complete(&err);
         return;
    }
 }
```

Thanks to this support added for horovod: https://github.com/apache/incubator-mxnet/pull/13932 we may be able to leverage this to call async_on_complete with the error.

Reply:
Adding wait_to_read in custom op can solve 1&2, and PushSync can be removed since the computation is already finished.

Reply:
we probably still need PushSync for the Sparse ndarray updates.

Reply:
we still need ExecType::kAsync. Custom operator is still async and when push is called it just pushes it into its custom op worker queue for execution later. Async will ensure that the threaded_engine_pooled and threaded_engine_per_device treat it as a special case and execute immediately instead of pushing the work again to one of the engine worker thread queue. Pushing to engine worker thread queue is unnecessary for custom op.

Reply:
After testing, ExecType::kAsync is really needed. Adding wait_to_read in engine worker thread will cause deadlock.
But PushSync can be removed and works well.

Reply:
we probably still need it for sparse. since for sparse we are updating chunk it is a write option. WaitToRead may not be enough.

Reply:
Yes, I also add WaitToWrite to make sure there's no left out exceptions.